mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryQuotaThresholdChangesStore.java

```diff
@@ -0,0 +1,83 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.inmemory.quota;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.IntStream;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.quota.QuotaThresholdChangesStore;
+import org.apache.james.mailbox.quota.model.QuotaThresholdChange;
+import org.apache.james.mailbox.quota.model.QuotaThresholdChanges;
+
+public class InMemoryQuotaThresholdChangesStore implements QuotaThresholdChangesStore {
+    private static final int MAX_RETRY = 50;
+
+    private final ConcurrentHashMap<User, QuotaThresholdChanges> countChanges;
+    private final ConcurrentHashMap<User, QuotaThresholdChanges> sizeChanges;
+
+    public InMemoryQuotaThresholdChangesStore() {
+        countChanges = new ConcurrentHashMap<>();
+        sizeChanges = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public QuotaThresholdChanges retrieveQuotaSizeThresholdChanges(User user) {
+        return Optional.ofNullable(sizeChanges.get(user))
+            .orElse(new QuotaThresholdChanges());
+    }
+
+    @Override
+    public void persistQuotaSizeThresholdChange(User user, QuotaThresholdChange change) {
+        doAddChange(sizeChanges, user, change);
+    }
+
+    @Override
+    public QuotaThresholdChanges retrieveQuotaCountThresholdChanges(User user) {
+        return Optional.ofNullable(countChanges.get(user))
+            .orElse(new QuotaThresholdChanges());    }
+
+    @Override
+    public void persistQuotaCountThresholdChange(User user, QuotaThresholdChange change) {
+        doAddChange(countChanges, user, change);
+    }
+
+    private void doAddChange(ConcurrentHashMap<User, QuotaThresholdChanges> map, User user, QuotaThresholdChange change) {
+        IntStream.range(0, MAX_RETRY)
+            .filter(i -> tryAddChange(map, user, change))
```

Comment:
it's not the right place to retry because the decision has already be taken elsewhere. In case of failure, you should redo the whole process with last version of history

Reply:
And you need to carry the version of the history you had to take the decision

Reply:
Concerning retrying with updated version of the history, it's not needed.

We are storing separate changes, and order don't matter: we are sorting the history upon retrieving it. Thus, we can have re-orders, it does not matter.

I'll add tests for out of order updates.

Reply:
You also said during a daily that you are trying to do "Event Sourcing" or somehow try to be close to it. In Event Sourcing, the store can't make assumption about the constraint the Aggregate need to take a decision, it's why we require full ordering.
What do you want to do actually ?

Reply:
Retrieve the list of changes then sorts it to enforce order, as performed in later commits.

(I'm expecting the history to be little, and in case of problems, we can easily prune it.)